### DISCUSSION TITLE
How do the times in `t_eval` or in my experiment relate to the solver steps?

### DISCUSSION THREAD

**Discussion by rtimms** [pybamm-team/PyBaMM#2044] [Q&A] [answered]
When I solve a model how do the times I request in `t_eval` (or the times set by the period of my experiment) relate to the timesteps taken by the solver?

**Comment by rtimms** [comment, marked as answer]
Internally the solvers use adaptive timestepping to meet the tolerances (`atol` and `rtol`) passed to the solver. The times in `t_eval` are the times at which you would like the solver to return a solution. The solver takes the steps it needs to in order to meet the error tolerances, but then gives you back a solution at the times you request (which may be different from the steps the solver actually took). 

There is a nice description of this in some [usage notes](https://computing.llnl.gov/projects/sundials/usage-notes) on the SUNDIALS website. You can think of the times in `t_eval` as the times in `tout` in the "normal" mode described in the linked usage notes.

**Reply by ljt999** [reply]
Hi Dr. Robert Timms,

I have reviewed the casadi_solver and solution files, but I was unable to find the specifics of how PyBaMM performs interpolation to obtain values at the t_eval points. Could you possibly pinpoint the section of the code where this process takes place? I greatly appreciate your assistance and time in advance.

Best regards,
Yunlu
